### PR TITLE
Add check if the number of samples is greater or equal to the number of shards in readers

### PR DESCRIPTION
**Pull request by JanuszL**
- when the number of shards outweighs the number of samples in readers
  enabling pading causes segmentation fault (as some shards don't
  have a sample that can be duplicated for padding). Also it has a little
  sense to have more shards than samples. This PR adds a relevant
  enforce and a corresponding test

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Category:

**Bug fix** (*non-breaking change which fixes an issue*)

## Description:
- when the number of shards outweighs the number of samples in readers
  enabling pading causes segmentation fault (as some shards don't
  have a sample that can be duplicated for padding). Also it has a little
  sense to have more shards than samples. This PR adds a relevant
  enforce and a corresponding test

### Affected modules and functionalities:
- loaders/readers

### Key points relevant for the review:
- NA

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/loader/loader.h [hunk 1]
```diff
@@ -218,6 +218,9 @@ class Loader {
         PrepareMetadataImpl();
         std::atomic_thread_fence(std::memory_order_release);
         loading_flag_ = true;
+        DALI_ENFORCE(num_shards_ <= Size(), make_string("The number of input samples: ", Size(),
+                                        ", needs to be at least equal to the number of the shards:",
```
Comment (klecki) [inline dali/operators/reader/loader/loader.h:222]: Nitpick:
```suggestion
                                        ", needs to be at least equal to the requested number of shards: ",
```

**Review by klecki — APPROVED** [review]

**Comment by JanuszL** [comment]
!build

**Review by klecki — APPROVED** [review]

**Comment by JanuszL** [comment]
!build

**Review by awolant — APPROVED** [review]

**Pull request merged**